x/exchange/orders_test.go

```diff
@@ -0,0 +1,2786 @@
+		},
+		{
+			name:     "unknown order type",
+			order:    newUnknownOrder(4),
+			expected: "*exchange.unknownOrderType",
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual string
+			testFunc := func() {
+				actual = tc.order.GetOrderType()
+			}
+			require.NotPanics(t, testFunc, "GetOrderType()")
+			assert.Equal(t, tc.expected, actual, "GetOrderType() result")
+		})
+	}
+}
+
+func TestOrder_GetOrderTypeByte(t *testing.T) {
+	tests := []struct {
+		name     string
+		order    *Order
+		expected byte
+		expPanic string
+	}{
+		{
+			name:     "AskOrder",
+			order:    NewOrder(1).WithAsk(&AskOrder{}),
+			expected: OrderTypeByteAsk,
+		},
+		{
+			name:     "BidOrder",
+			order:    NewOrder(2).WithBid(&BidOrder{}),
+			expected: OrderTypeByteBid,
+		},
+		{
+			name:     "nil inside order",
+			order:    NewOrder(3),
+			expPanic: nilSubTypeErr(3),
+		},
+		{
+			name:     "unknown order type",
+			order:    newUnknownOrder(4),
+			expPanic: unknownSubTypeErr(4),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual byte
+			testFunc := func() {
+				actual = tc.order.GetOrderTypeByte()
+			}
+			assertions.RequirePanicEquals(t, testFunc, tc.expPanic, "GetOrderTypeByte()")
+			assert.Equal(t, tc.expected, actual, "GetOrderTypeByte() result")
+		})
+	}
+}
+
+func TestOrder_GetHoldAmount(t *testing.T) {
+	tests := []struct {
+		name     string
+		order    *Order
+		expected sdk.Coins
+		expPanic string
+	}{
+		{
+			name: "AskOrder",
+			order: NewOrder(1).WithAsk(&AskOrder{
+				Assets:                  sdk.NewInt64Coin("acorns", 85),
+				SellerSettlementFlatFee: &sdk.Coin{Denom: "bananas", Amount: sdkmath.NewInt(12)},
+			}),
+			expected: sdk.NewCoins(sdk.NewInt64Coin("acorns", 85), sdk.NewInt64Coin("bananas", 12)),
+		},
+		{
+			name: "BidOrder",
+			order: NewOrder(2).WithBid(&BidOrder{
+				Price:               sdk.NewInt64Coin("acorns", 85),
+				BuyerSettlementFees: sdk.NewCoins(sdk.NewInt64Coin("bananas", 4), sdk.NewInt64Coin("cucumber", 7)),
+			}),
+			expected: sdk.NewCoins(
+				sdk.NewInt64Coin("acorns", 85),
+				sdk.NewInt64Coin("bananas", 4),
+				sdk.NewInt64Coin("cucumber", 7),
+			),
+		},
+		{
+			name:     "nil inside order",
+			order:    NewOrder(3),
+			expPanic: nilSubTypeErr(3),
+		},
+		{
+			name:     "unknown order type",
+			order:    newUnknownOrder(4),
+			expPanic: unknownSubTypeErr(4),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual sdk.Coins
+			testFunc := func() {
+				actual = tc.order.GetHoldAmount()
+			}
+			assertions.RequirePanicEquals(t, testFunc, tc.expPanic, "GetHoldAmount()")
+			assert.Equal(t, tc.expected.String(), actual.String(), "GetHoldAmount() result")
+		})
+	}
+}
+
+func TestOrder_Validate(t *testing.T) {
+	// Annoyingly, sdkmath.Int{} (the zero value) causes panic whenever you
+	// try to do anything with it. So we have to give it something.
```

Comment:
This is nice to know, but probably isn't needed since it was fixed.

Reply:
`sdkmath.Int{}.Equal(sdkmath.Zero())` will panic, as will `sdkmath.ZeroInt().Equal(sdkmath.Int{})`. Same with all the comparison and math methods including `.IsZero()`, `.IsPositive()`, `.IsNegative()`, and `.Add(...)`.

Unrelatedly, but also annoying as hell, `coinA.IsEqual(coinB)` will also panic if the two coins have different denoms (regardless of the amounts). Yay! And `coinsA.IsEqual(sdk.Coin{Denom: <same as A>})` will also panic because of the empty int comparison.